dnadoh/synthesize.py

```diff
@@ -0,0 +1,335 @@
+
+
+def _write(output_stem, genomes, people):
+    """Write people as CSV."""
+    assert people
+    _write_overall(output_stem, people)
+    _write_reference_genome(output_stem, genomes)
+    _write_variants(output_stem, genomes, people)
+    _write_people(output_stem, people)
+
+
+def _write_overall(output_stem, people):
+    """Write DNA sequences and people for reference."""
+    filename = f"{output_stem}-overall.csv"
+    headings = people[0].dict().keys()
+    with open(filename, "w") as raw:
+        writer = csv.DictWriter(raw, fieldnames=headings)
+        writer.writeheader()
+        for person in people:
+            writer.writerow(person.dict())
+
+
+def _write_reference_genome(output_stem, genomes):
+    """Save the reference genome and related information."""
+    filename = f"{output_stem}-reference.json"
+    data = {"genome": genomes.reference, "locations": list(sorted(genomes.locations))}
+    with open(filename, "w") as writer:
+        json.dump(data, writer)
+
+
+def _write_variants(output_stem, genomes, people):
+    """Write one variant file per person."""
+    width = len(str(len(people)))
+    for (pid, person) in enumerate(people):
+        pid_str = str(pid).zfill(width)
+        filename = f"{output_stem}-pid{pid_str}.csv"
+        with open(filename, "w") as raw:
+            writer = csv.DictWriter(raw, fieldnames=["loc", "base"])
+            writer.writeheader()
+            for i in range(len(genomes.reference)):
+                if person.genome[i] != genomes.reference[i]:
+                    writer.writerow({"loc": i, "base": person.genome[i]})
+
+
+def _write_people(output_stem, people):
+    """Write phenotypic data for people."""
+    filename = f"{output_stem}-people.csv"
+    headings = people[0].dict()
+    del headings["genome"]
+    headings["pid"] = 0
+    with open(filename, "w") as raw:
+        writer = csv.DictWriter(raw, fieldnames=headings)
+        writer.writeheader()
+        for (pid, person) in enumerate(people):
+            details = person.dict()
+            del details["genome"]
+            details["pid"] = pid
+            writer.writerow(details)
+
+
+# --------------------------------------------------------------------------------------
+# Utilities
+# --------------------------------------------------------------------------------------
+
+
+def _other_bases(seq, loc):
+    """Create a list of bases minus the one in the sequence at that location."""
+    return list(set(DNA) - {seq[loc]})
+
+
+def _truncate(num, digits):
+    """Truncate a number to the specified number of fractional digits."""
+    scale = 10**digits
+    return int(scale * num) / scale
+
+
+# --------------------------------------------------------------------------------------
+# Main driver
+# --------------------------------------------------------------------------------------
+
+
+def main():
+    """Main driver."""
+    options = parse_args()
+    random.seed(options.seed)
```

Comment:
can we set a default seed?

Reply:
Easily, but it's generally regarded as a bad idea: it makes it very easy for people to all generate the same results without realizing it. Happy to chat...

Reply:
Can you check if the code runs if `options.seed` is not set on the command line when running?